src/core/config/merge.js

```diff
@@ -15,25 +15,38 @@ import deepExtend from "deep-extend"
 
 const merge = (target, ...sources) => {
   let domNode = Symbol.for("domNode")
-  const sourcesWithoutDomNode = []
+  let primaryName = null
+  const sourcesWithoutExceptions = []
 
   for (const source of sources) {
+    const sourceWithoutExceptions = { ...source }
+
     if (Object.hasOwn(source, "domNode")) {
       domNode = source.domNode
-      const sourceWithoutDomNode = { ...source }
-      delete sourceWithoutDomNode.domNode
-      sourcesWithoutDomNode.push(sourceWithoutDomNode)
-    } else {
-      sourcesWithoutDomNode.push(source)
+      delete sourceWithoutExceptions.domNode
+    }
+
+    if (source["urls.primaryName"]) {
+      primaryName = source["urls.primaryName"]
+      delete sourceWithoutExceptions["urls.primaryName"]
+    } else if (source.urls && source.urls.primaryName) {
+      primaryName = source.urls.primaryName
+      delete sourceWithoutExceptions.urls.primaryName
     }
+
+    sourcesWithoutExceptions.push(sourceWithoutExceptions)
   }
 
-  const merged = deepExtend(target, ...sourcesWithoutDomNode)
+  const merged = deepExtend(target, ...sourcesWithoutExceptions)
 
   if (domNode !== Symbol.for("domNode")) {
     merged.domNode = domNode
   }
 
+  if (primaryName) {
+    merged.urls.primaryName = primaryName
```

Comment:
We have no way of knowing if `merged.urls` exists at this point. It's because `urls.primaryName` could have just come from `queryOptions`, but nobody bothered to set the `urls` option.

```suggestion
if (primaryName !== Symbol.for('urls.primaryName') && Array.isArray(merged.urls)) {
    merged.urls.primaryName = primaryName
}
```